### PR TITLE
Implement history timeline API (backend)

### LINKED ISSUE #4424: API for Deployment History of an Instance
Request to have a single API endpoint, e.g. `api/v1/projects/{instanceId}/history` that returns details on every deployment made to that instance in the context of version history.

If I'm not mistaken, the sources for these deployments would be:

- Node-RED Editor - user has clicked the "Deploy" button themselves
- DevOps Pipeline - where we would want detail on where the flows came from
- Rolback Snapshots

A lot of these I think could be gathered from the Audit Log (and reformatted appropriately). 

It is also important to verify if each of these steps in the `history` have an available snapshot that we can revert to. Part of this work is to also determine if there are actually new tables or events we need to be recording in order to support this functionality.

### PR DESCRIPTION
**Pull request by Steve-Mcl**
closes #4424 

## Description

* Refactor preHander in `forge/routes/api/project.js` so it can be used by EE and NON EE routes
* Enable snapshot audit logging for "auto snapshots"
* Add `forProjectHistory` to `forge/db/models/AuditLog.js` for querying audit logs for a projects timeline
* Add Schema and Views in `forge/db/views/AuditLog.js`
* Improve output for audit snapshot entries object to include the `description` field
* Add EE permission `project:history`
* Add route `GET /projects/:projectId/history` in the EE section `forge/ee/routes/projectHistory/index.js`
* Add team type feature flag `projectHistory`
* Add Unit tests

####  `test/unit/forge/ee/routes/api/project_spec.js`
```
  Project (EE)
    Project History
      ✔ Owner should get a timeline of changes to the project
      ✔ Member should get a timeline of changes to the project
      ✔ Viewer should not be able to access project history (403)
      ✔ Non member should not be able to access project history (404)
      ✔ Anonymous should not be able to access project history (401)
      ✔ Should return 404 when the teamtype feature flag is disabled
      Timeline Data
        ✔ Should return a timeline of changes to the project
        Pagination
          ✔ Should limit response
          ✔ Should use cursor
```

#### `test/unit/forge/routes/api/project_spec.js`
```
  Project API
    Project History
      ✔ Should not get a timeline of changes to the project (18261ms)
```

### TODO:

- [x] feature flag

## Related Issue(s)

#4424 

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [x] Suitable unit/system level tests have been added and they pass 
 - [ ] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

File: test/unit/forge/ee/routes/api/project_spec.js [hunk 1]
```diff
@@ -266,3 +271,334 @@ describe('npmrc values should be hidden from none owners', function () {
...
+                const defaultTeamType = await app.db.models.TeamType.findOne({ where: { name: 'starter' } })
+                const defaultTeamTypeProperties = defaultTeamType.properties
+                defaultTeamTypeProperties.features.instanceAutoSnapshot = true
+                defaultTeamType.properties = defaultTeamTypeProperties
+                await defaultTeamType.save()
+
+                // Simulate below events by pushing entries to the audit log and generating snapshots
+                // 1. Create snapshot 1
+                // 2. Modify flows
+                // 3. Create snapshot 2
```
Comment (Steve-Mcl) [inline test/unit/forge/ee/routes/api/project_spec.js:490]:
```suggestion
                // 3. Create snapshot 2 (auto snapshot)
```

File: forge/db/models/AuditLog.js [hunk 2]
```diff
@@ -96,6 +96,108 @@ module.exports = {
...
+                },
+                forProjectHistory: async (projectId, pagination = {}) => {
+                    // Premise:
+                    //  we want to generate a timeline of events for a project, including snapshots
+                    //  so that user can see "things that changed" the project and any immediate snapshots.
+                    // Approach:
+                    //  1. Get all log entries for the project starting from the pagination cursor & limited by the pagination limit
+                    //     where they meet the criteria for project history (see Op.in filter below)
+                    //  2. If the log entry has a snapshot, try to find the actual snapshot object and replace the body with it
+                    //     This gives updates any stale snapshot references in the log entries
```
Comment (Steve-Mcl) [inline forge/db/models/AuditLog.js:108]:
```suggestion
                    //     This updates any stale snapshot references in the log entries
```

File: forge/db/models/AuditLog.js [hunk 3]
```diff
@@ -96,6 +96,108 @@ module.exports = {
...
+                    //     Additionally, flag snapshot existence in the info object as { snapshotExists: true/false }
+                    //     (The info object is a permitted field in the log entry body)
```
Comment (Steve-Mcl) [inline forge/db/models/AuditLog.js:110]:
```suggestion
                    //     (The info object is a permitted field in the audit log entry body)
```

clarify

**Comment by Steve-Mcl** [comment]
@cstns some sample data from the API for your consideration when working on the front-end...

<details><summary> A request for 5 items, no cursor</summary>
<p>

`GET /api/v1/projects/4d1381e8-67e7-40ec-b39e-07ac159df3c4/history?limit=5`

```json
{
  "meta": {
    "next_cursor": "2aDOkWmDMb"
  },
  "count": 5,
  "timeline": [
    {
      "id": "MqDZqRA8P7",
      "createdAt": "2024-09-18T11:58:29.054Z",
      "user": {
        "id": "system",
        "username": "System",
        "name": "FlowFuse Platform",
        "avatar": "/avatar/camera.svg"
      },
      "event": "project.snapshot.created",
      "data": {
        "snapshot": {
          "id": "J5yD3kEDr9",
          "name": "Auto Snapshot - 2024-09-18 12:58:28",
          "description": "Instance Auto Snapshot taken following a Modified Nodes deployment"
        },
        "info": {
          "snapshotExists": true
        }
      }
    },
    {
      "id": "qVYl7ZND7p",
      "createdAt": "2024-09-18T11:58:28.966Z",
      "user": {
        "id": "EexY4j17B2",
        "username": "steve",
        "name": "steve-mcl",
        "avatar": "http://192.168.86.45:3000/avatar/c3RldmU"
      },
      "event": "flows.set",
      "data": {
        "flowsSet": {
          "type": "nodes"
        }
      }
    },
    {
      "id": "WjYXGWRD3d",
      "createdAt": "2024-09-17T18:55:20.099Z",
      "user": {
        "id": "EexY4j17B2",
        "username": "steve",
        "name": "steve-mcl",
        "avatar": "http://192.168.86.45:3000/avatar/c3RldmU"
      },
      "event": "project.snapshot.imported",
      "data": {
        "sourceProject": {
          "id": "178e4d8a-4075-4647-9e62-8fe3eba65dec",
          "name": "prod"
        },
        "snapshot": {
          "id": "x17DdQpgKb",
          "name": "a copy of dashboard soc - Deploy Snapshot - 2024-09-17 19:55:17",
          "description": "Snapshot created for pipeline deployment from src to dst as part of pipeline test"
        },
        "info": {
          "snapshotExists": true
        }
      }
    },
    {
      "id": "dz8p43P8xK",
      "createdAt": "2024-09-17T18:51:29.307Z",
      "user": {
        "id": "EexY4j17B2",
        "username": "steve",
        "name": "steve-mcl",
        "avatar": "http://192.168.86.45:3000/avatar/c3RldmU"
      },
      "event": "project.settings.updated",
      "data": {
        "updates": [
          {
            "key": "header.title",
            "old": "dashboard-soc",
            "new": "dashboard-soc2",
            "dif": "updated"
          }
        ]
      }
    },
    {
      "id": "2aDOkWmDMb",
      "createdAt": "2024-09-17T18:48:37.803Z",
      "user": {
        "id": "EexY4j17B2",
        "username": "steve",
        "name": "steve-mcl",
        "avatar": "http://192.168.86.45:3000/avatar/c3RldmU"
      },
      "event": "project.snapshot.rolled-back",
      "data": {
        "snapshot": {
          "id": "Q2YDLwlO9M",
          "name": "new 1",
          "description": ""
        },
        "info": {
          "snapshotExists": true
        }
      }
    }
  ]
}
```

</p>
</details> 


<details><summary>A request for 100, but the cursor is near the end of the data</summary>
<p>

`GET /api/v1/projects/4d1381e8-67e7-40ec-b39e-07ac159df3c4/history?limit=100&cursor=REYvEy5DLa`

```json
{
  "meta": {},
  "count": 2,
  "timeline": [
    {
      "id": "MqDZ75AYP7",
      "createdAt": "2024-07-06T06:36:34.483Z",
      "user": {
        "id": "EexY4j17B2",
        "username": "steve",
        "name": "steve-mcl",
        "avatar": "http://192.168.86.45:3000/avatar/c3RldmU"
      },
      "event": "flows.set",
      "data": {
        "flowsSet": {
          "type": "full"
        }
      }
    },
    {
      "id": "568ryyAgO3",
      "createdAt": "2024-05-30T10:55:27.170Z",
      "user": {
        "id": "EexY4j17B2",
        "username": "steve",
        "name": "steve-mcl",
        "avatar": "http://192.168.86.45:3000/avatar/c3RldmU"
      },
      "event": "project.created",
      "data": {
        "team": {
          "id": "2VJ1bDYopr",
          "name": "Team team",
          "slug": "team1",
          "type": null
        }
      }
    }
  ]
}
```

</p>
</details>

**Review by knolleary — CHANGES_REQUESTED** [review]
Some feedback on the implementation around scalability of db queries.

But I have some more fundamental questions around why a separate API is needed above the existing `/audit-log` end point which could be extended to include the additional meta information - https://github.com/FlowFuse/flowfuse/issues/4424#issuecomment-2360557072

Maybe a quick chat would be helpful to clarify this.

**Comment by Steve-Mcl** [comment]
discussion notes:
* mark this API private/internal to prevent lock-in
* ensure snapshot flows and any data not needed in the snapshot is not included
* address scalability issue

File: forge/db/models/AuditLog.js [hunk 4]
```diff
@@ -96,6 +96,108 @@ module.exports = {
...
+                            entry.body = ((typeof entry.body || '{}') === 'string' ? JSON.parse(entry.body) : entry.body) || {}
+                            // since we are in the context of a project history, we don't need to include the project object (redundant)
+                            delete entry.body.project
+                            // update snapshot if available and flag existence
+                            if (entry.body.snapshot) {
+                                if (typeof entry.body.info !== 'object') {
+                                    entry.body.info = entry.body.info ? { _info: entry.body.info } : {}
+                                }
+                                if (entry.body.snapshot) {
+                                    const snapshot = await M.ProjectSnapshot.byId(entry.body.snapshot.id)
```
[the remark this answers is not in the document]
Comment (Steve-Mcl) [reply forge/db/models/AuditLog.js]: resolved in:
* [query snapshots in one go](https://github.com/FlowFuse/flowfuse/pull/4509/commits/629659102d9c92351da9a27f872780b0abf7d9fc)  [6296591](https://github.com/FlowFuse/flowfuse/pull/4509/commits/629659102d9c92351da9a27f872780b0abf7d9fc)

* [mark route as internal](https://github.com/FlowFuse/flowfuse/pull/4509/commits/4f6f14ed5d57811d9ad65b64ca4f88d1f8ba103b) [4f6f14e](https://github.com/FlowFuse/flowfuse/pull/4509/commits/4f6f14ed5d57811d9ad65b64ca4f88d1f8ba103b)

**Comment by Steve-Mcl** [comment]
> * mark this API private/internal to prevent lock-in

addressed in [mark route as internal](https://github.com/FlowFuse/flowfuse/pull/4509/commits/4f6f14ed5d57811d9ad65b64ca4f88d1f8ba103b)  [4f6f14e](https://github.com/FlowFuse/flowfuse/pull/4509/commits/4f6f14ed5d57811d9ad65b64ca4f88d1f8ba103b)



> * ensure snapshot flows and any data not needed in the snapshot is not included
> * address scalability issue

resolved in [query snapshots in one go](https://github.com/FlowFuse/flowfuse/pull/4509/commits/629659102d9c92351da9a27f872780b0abf7d9fc) [6296591](https://github.com/FlowFuse/flowfuse/pull/4509/commits/629659102d9c92351da9a27f872780b0abf7d9fc)

**Comment by Steve-Mcl** [comment]
@knolleary after much ado, the unit tests are now passing.

FYI, The test runner was too fast and audit entries had the same timestamp causing the sort ordering to go by ID ASC when 2 timestamps were the same.

We can leave the API returning results by ID DESC or introduce small delays in the test runner to simulate more realistic operational logging.  What do you think?

**Comment by hardillb** [comment]
I vote add delay to test, but are we 100% that it's impossible to go that fast in real life?

**Comment by knolleary** [comment]
Real-life doesn't matter here; two events happening in the same millisecond won't matter to the user what order they are reported in the log.

The test however is making assumptions about the order. So either guarantee the order or make the test more flexible to cope with the variable ordering

**Review by knolleary — APPROVED** [review]

**Comment by Steve-Mcl** [comment]
> @Steve-Mcl if a flow is set via a DevOps Pipeline, what event lists in this response?

  'project.created',
  'project.deleted',
  'flows.set', // flows deployed by user
  'project.settings.updated',
  'project.snapshot.created', // snapshot created manually or automatically
  'project.snapshot.rolled-back', // snapshot rolled back by user
  'project.snapshot.imported' // result of a pipeline deployment

https://github.com/FlowFuse/flowfuse/pull/4509/files#diff-bf6b77295e5a9729998a556bd9047d8357ccf976f9a052f514abf7f148b3b781R119-R125

**Pull request merged**